File: docs/js/templates.js
// Mirror of templates.py — same prompt template, same vocabulary, same order.
// The template is the interface to the model: a sentence with a hole in it.

export const TAG_TEMPLATE = 'a photo of a {tag}';
export const CAPTION_TEMPLATE = 'a photo of {tags}';

// A sentence with NO tag in it — the baseline multi-label scoring
// (labels.js) compares every tag prompt against.
export const NEUTRAL_PROMPT = 'a photo';

// The pool the embedding agent (agent.js) draws proposals from, best first.
export const TEMPLATE_POOL = [
  TAG_TEMPLATE,
  'a close-up photo of a {tag}',
  'a photograph of a {tag}',
  'an image showing a {tag}',
];

export const VOCAB = ["cat","dog","bear","parrot","bird","horse","fish","insect",
  "animal","pet","wildlife","waterfall","mountain","forest","beach","lake",
  "river","sky","sunset","snow","flower","sunflower","tree","garden",
  "landscape","city","street","tower","castle","palace","church","bridge",
  "building","landmark","architecture","house","car","bicycle","train",
  "airplane","boat","vehicle","pizza","apple","strawberry","fruit","food",
  "drink","dessert","person","portrait","crowd","planet","moon","stars",
  "space","painting","toy","book","computer"];

// One ready-to-embed sentence per candidate tag (templates.tag_prompts).
export const tagPrompts = (template = TAG_TEMPLATE) =>
  VOCAB.map(t => template.replace('{tag}', t));

// Build a caption sentence out of the winning tags (templates.caption_for).
export const captionFor = tags => CAPTION_TEMPLATE.replace('{tags}', tags.join(', '));

File: templates.py
"""Prompt templates: how we phrase questions to a vision-language model.

CLIP was trained on (image, caption) pairs, so it understands images best
through natural-language sentences. A "prompt template" is just a sentence
with a hole in it — we fill the hole with a candidate tag and ask CLIP
"how well does this sentence describe this image?".
"""

# The classic zero-shot template from the CLIP paper.
DEFAULT_TAG_TEMPLATE = "a photo of a {tag}"

# How we turn the winning tags into a stored caption.
DEFAULT_CAPTION_TEMPLATE = "a photo of {tags}"

# Candidate meta tags. Add your own words here — that is the whole "training".
TAG_VOCABULARY = [
    "cat", "dog", "bear", "parrot", "bird", "horse", "fish", "insect",
    "animal", "pet", "wildlife",
    "waterfall", "mountain", "forest", "beach", "lake", "river", "sky",
    "sunset", "snow", "flower", "sunflower", "tree", "garden", "landscape",
    "city", "street", "tower", "castle", "palace", "church", "bridge",
    "building", "landmark", "architecture", "house",
    "car", "bicycle", "train", "airplane", "boat", "vehicle",
    "pizza", "apple", "strawberry", "fruit", "food", "drink", "dessert",
    "person", "portrait", "crowd",
    "planet", "moon", "stars", "space",
    "painting", "toy", "book", "computer",
]


def fill(template: str, **values) -> str:
    """Fill the {holes} in a template. That's all a prompt template is."""
    return template.format(**values)


def tag_prompts(template: str = DEFAULT_TAG_TEMPLATE, vocabulary=TAG_VOCABULARY):
    """One ready-to-embed sentence per candidate tag."""
    return [fill(template, tag=tag) for tag in vocabulary]


def caption_for(tags, template: str = DEFAULT_CAPTION_TEMPLATE) -> str:
    """Build a caption sentence out of the winning tags."""
    return fill(template, tags=", ".join(tags))
